timeseries plugin

jburrow opened the pull request:
Hi pyatyispyatil,

I just wanted to give you visibility of my fork

- I have converted the examples to typescript
- enforced prettier on the whole code-base (including examples and config outside of /src/
- introduced a new plugin for rendering time-series data
- refactored to the example to have explicit usage there there it is explicit which plugins are configured an enabled
- demonstrated the use of multiple instances of the same plugin type.
- I have struggled to .ts convert the test-data code fully - i have some typings and it is better - but therea re still some any[] there - which i haven't fully typed.

I can spend a few ours changing this to your liking...


James

![image](https://user-images.githubusercontent.com/1444266/209701052-91849e60-ddef-4eb4-99af-d7c53678b580.png)

Smrtnyk commented on tsconfig.json:
```diff
@@ -1,110 +1,12 @@
-  "exclude": [
-    "./src/**/*.spec.ts"
-  ]
+    "compilerOptions": {
```
I find these comments descriptive, and it shows what we are still missing in order to be fully strict
I would actually prefer to restore the changes on this file
also target changed to 2019? Any reason for that, are you using older browser?

jburrow replied:
2019 is the last version that webpack 4.0 support - I know this is bit lame; But we have a complex build pipeline on one of my projects, and we don't want the overhead of transpiling node_modules to support es2019+

If you are happy for the downgrade - that would be handy for my use-case.

RE: The tsconfig purge.
When converting to the examples - I couldn't figure our what was significant and what wasn't. I went back to the default (basically empty) then built it back up....

There is a bit more effort to get the project fully strict ... i did try to fix a few extras too.

pyatyispyatil replied:
I think it's not so scary to slightly improve compatibility and change the target to 2019. And at the expense of strict mode - I agree with @Smrtnyk. Perhaps we should leave only those options that do not allow us to enable strict mode?

Smrtnyk commented:
It would be nicer if you could separate prettier changes to its own PR
that would make timeseries plugin isolated to its own PR and easier to review

jburrow commented:
> It would be nicer if you could separate prettier changes to its own PR that would make timeseries plugin isolated to its own PR and easier to review

Agreed; This has conflated into to heaps of things....

I started off breaking off branches & PRs - It just has an overhead maintaining them.

pyatyispyatil commented:
Man, very cool pr, but unfortunately it seems that the feature is a little incomplete - in chrome under macOS, something strange happened to the flame chart for some reason.
<img width="1440" alt="image" src="https://user-images.githubusercontent.com/4976306/209724787-a6dd4312-2e56-4e22-b562-ce045291eef0.png">

Smrtnyk commented:
> > It would be nicer if you could separate prettier changes to its own PR that would make timeseries plugin isolated to its own PR and easier to review
> 
> Agreed; This has conflated into to heaps of things....
> 
> I started off breaking off branches & PRs - It just has an overhead maintaining them.

yeah I think it could be separated as following:
PR for example conversion to TS
PR for prettier changes
PR for timeseries plugin

pyatyispyatil commented:
When scrolling, a glitch is noticeable: the extreme points seem to be attached to the edge of the screen, and because of this, the graph seems to change slightly. Apparently, you need to render part of the graph behind the screen (last and first points).

pyatyispyatil commented:
https://github.com/pyatyispyatil/flame-chart-js/pull/44#issuecomment-1366194472 This happens if the screen is not very large. And if you try to move the toggle plugin, then for some reason more additional plugins will be resized.

jburrow commented:
> When scrolling, a glitch is noticeable: the extreme points seem to be attached to the edge of the screen, and because of this, the graph seems to change slightly. Apparently, you need to render part of the graph behind the screen (last and first points).

thank-you - will try and fix - and let you know.

jburrow commented:
> > > It would be nicer if you could separate prettier changes to its own PR that would make timeseries plugin isolated to its own PR and easier to review
> > 
> > 
> > Agreed; This has conflated into to heaps of things....
> > I started off breaking off branches & PRs - It just has an overhead maintaining them.
> 
> yeah I think it could be separated as following: PR for example conversion to TS PR for prettier changes PR for timeseries

> Man, very cool pr, but unfortunately it seems that the feature is a little incomplete - in chrome under macOS, something strange happened to the flame chart for some reason. <img alt="image" width="1440" src="https://user-images.githubusercontent.com/4976306/209724787-a6dd4312-2e56-4e22-b562-ce045291eef0.png">

let me dig.

pyatyispyatil commented:
@jburrow It looks like most of the code style updates are already in the master. Can you update this PR so that we can start reviewing it?)

Smrtnyk commented on package.json:
```diff
@@ -48,19 +48,21 @@
     "rollup-plugin-node-builtins": "^2.1.2",
     "rollup-plugin-node-resolve": "^5.2.0",
-    "typescript": "4.8.4"
+    "ts-jest": "^29.0.3",
```
whats the reason of introducing a much slower ts-jest over swc/jest?

jburrow replied:
No good reason - other than familiarity - i had issues with the tests not running - so just switched over to ts-jest. That said - for 2 test scripts; the overhead comparision witll be fairly trivial.

I will remove this and re-check if the testing issue still exists

jburrow commented:
> @jburrow It looks like most of the code style updates are already in the master. Can you update this PR so that we can start reviewing it?)

I have thrown away most of the stuff i had and gone back to master.

pyatyispyatil commented:
Overall, it looks good, but it needs to be finished a little more.)
1. The code is dirty, with a lot of commented-out code - this should not be.
2. It is better to leave the default flame chart in the example that was before this pull request and add your new plugin to it.
3. The visual part does not look very good - when scrolling, the extreme value disappears and an empty area is obtained on the chart. 
4. A little earlier in this pull request there was a broken line instead of a stepped graph - maybe do both and add the ability to configure this in the parameters? It would also be cool to add smooth lines, but this can be added later.
5. For some reason, when hovering, I don’t get a tooltip, although I see that you described it in the code.
6. It would be very cool to be able to render several charts in one timeseries plugin and pass the color not as the second argument of the plugin constructor, but in the data itself.

<img width="529" alt="image" src="https://user-images.githubusercontent.com/4976306/210250695-c09f9e2d-696b-4ca7-a30c-fe598492dbd0.png">

jburrow commented:
Agreed. 100%

Will work thru this points shortly.

Smrtnyk commented on tsconfig.json:
```diff
@@ -65,7 +65,7 @@
         // "stripInternal": true,                            /* Disable emitting declarations that have `@internal` in their JSDoc comments. */
         // "noEmitHelpers": true,                            /* Disable generating custom helper functions like `__extends` in compiled output. */
-        "noEmitOnError": true /* Disable emitting files if any type checking errors are reported. */,
+        "noEmitOnError": false /* Disable emitting files if any type checking errors are reported. */,
```
Why switch to false?

jburrow replied:
I have fixed this in a seperate PR - the problem was this feature meant the hot-reload local dev server would exit if there was a typescript error requiring a restart; I have fixed this and sent a PR with a fix for that.

Smrtnyk commented on src/plugins/timeseries-plugin.ts:
```diff
@@ -0,0 +1,233 @@
+
+        const d: TimeseriesPointWithIndex[] = [];
+
+        this.data.forEach(([ts, v], idx) => {
```
can we use more descriptive names instead of `d`, `v`, `ts` ...
its hard to follow

jburrow replied:
sure - will rename - wasn't sure if an {} is better than a tuple in this case tbh.

jburrow commented:
> Overall, it looks good, but it needs to be finished a little more.)
> 
> 1. The code is dirty, with a lot of commented-out code - this should not be.
**Agreed - This is just a draft right now - i will address before finalizing**
> 2. It is better to leave the default flame chart in the example that was before this pull request and add your new plugin to it.
**Agreed - But I think we could improve how the default behavior works; Either with a FlameChart Factory method can build a flame chart from data - that then leaving the class constructor to be super clean with just plugins passed in.

In addition - i wonder if the plugins should have "supports toggle" boolean - so that there isn't the boilerplate for the toggle behaviour - what do you think?**


> 3. The visual part does not look very good - when scrolling, the extreme value disappears and an empty area is obtained on the chart.
**Yes - i am now fixing this stuff up.**

> 4. A little earlier in this pull request there was a broken line instead of a stepped graph - maybe do both and add the ability to configure this in the parameters? It would also be cool to add smooth lines, but this can be added later.
**Yes - an option to toggle between the line-chart style would be good idea**

> 5. For some reason, when hovering, I don’t get a tooltip, although I see that you described it in the code.
** Yes - will fix - just modifying the internals of the render atm"

> 6. It would be very cool to be able to render several charts in one timeseries plugin and pass the color not as the second argument of the plugin constructor, but in the data itself.
** Maybe? - Why not just have multiple instances of the plugin? -- just same smae but different? **

> 
> <img alt="image" width="529" src="https://user-images.githubusercontent.com/4976306/210250695-c09f9e2d-696b-4ca7-a30c-fe598492dbd0.png">

jburrow commented:
I think it is ready for review - let me know

pyatyispyatil commented:
> I think it is ready for review - let me know @pyatyispyatil ty

It would be great if you resolved the conflicts first.)

Smrtnyk commented on package.json:
```diff
@@ -56,6 +56,7 @@
         "jest": "^29.1.2",
         "pre-commit": "^1.2.2",
         "prettier": "^2.6.1",
+        "query-string": "^8.1.0",
```
URL constructor also has a nice api to interact with query string
would be cool to use that, instead of adding a new dep

jburrow replied:
I not familiar with that - will google it.

Smrtnyk replied:
https://developer.mozilla.org/en-US/docs/Web/API/URL
its a pretty cool Web API

jburrow replied:
removed

Smrtnyk commented on src/plugins/timeseries-plugin.ts:
```diff
@@ -0,0 +1,259 @@
+
+    override render() {
+        if (this.data?.length === 0 || !this.summary) {
+            console.log('[render] no data - returned');
```
not sure if we should log stuff to the user browser and create noise
if we should, then `[render] no data - returned` is not descriptive enough to figure out on what is the issue

jburrow replied:
will happily remove the console.

jburrow replied:
done

Smrtnyk commented on tsconfig.json:
```diff
@@ -100,6 +100,5 @@
     },
     "types": ["color", "events"],
-    "include": ["./src/**/*.ts", "./example/src/**/*.ts"],
-    "exclude": ["./src/**/*.spec.ts"]
```
why are test files now included in the prod tsconfig?

jburrow replied:
because lint complained - that they were not part of the tsconfig project.

Smrtnyk replied:
won't that cause tsc to also compile test files when building for production?

jburrow replied:
yeah it does - ... looking

jburrow replied:
remvoed

Smrtnyk replied:
eslint was using tsconfig.eslint.json before, which includes spec files also
not sure what is the issue here, but it worked before and nothing complained

tsconfig.json was used for prod - excluded spec files
tsconfig.spec.json was used for spec files and only included spec files
tsconfig.eslint.json was used for eslint and included everything

which IDE are you using? and can you post the lint error you are facing?

Smrtnyk commented on src/plugins/timeseries-plugin.ts:
```diff
@@ -0,0 +1,259 @@
+type TimeseriesPointWithIndex = [idx: number, ts: number, value: number, normalizedTs: number, normalizedValue: number];
+
+enum TimeseriesPointWithIndexX {
+    idx = 0,
```
til now our enum members were always uppercased
not sure to what we should stick with, but it should be consistent

jburrow replied:
will happily change

jburrow replied:
done

Smrtnyk commented on src/plugins/timeseries-plugin.ts:
```diff
@@ -0,0 +1,259 @@
+
+enum TimeseriesPointWithIndexX {
+    idx = 0,
+    ts = 1,
```
could we rename `ts` to a proper word? 
I always struggle to a abbreviations if they are not common enough for me to recognize

jburrow replied:
will change to timestamp

jburrow replied:
done

jburrow commented on src/plugins/flame-chart-plugin.ts:
```diff
@@ -306,7 +306,7 @@ export class FlameChartPlugin extends UIPlugin {
             this.interactionsEngine.clearHitRegions();
             this.actualClusterizedFlatTree.forEach(processCluster(addHitRegion));
-        }, 16);
+        }, 16) as unknown as number;
```
It is because the typings are getting confused with the node builtins - which has a different return type. This just quieten's the linter

Smrtnyk replied:
We might need to adjust the lib field in tsconfig then, to add `DOM` into it
or use `window.setTimeout`

Smrtnyk replied:
usually we should avoid casts
but I am just giving opinion here
it is up to @pyatyispyatil to decide

jburrow replied:
For some reason it is linting against nodejs api - not browser apis... this is really no-op.
I am using vscode - and linting both in vscode - and via npm run lint on full project.

jburrow replied:
![image](https://user-images.githubusercontent.com/1444266/211146822-e31316eb-f210-48d5-b955-01dd1e0a4ab7.png)

pyatyispyatil replied:
<img width="931" alt="image" src="https://user-images.githubusercontent.com/4976306/211202060-2f0e907d-a7da-4b80-b380-12a5bda3d1f9.png">
I think that @Smrtnyk is absolutely right. You somehow need to figure out your ide.)

jburrow replied:
ty

jburrow commented:
> > I think it is ready for review - let me know @pyatyispyatil ty
> 
> It would be great if you resolve the conflicts first.)

done - i noticed you hadn't run "npm run pretty" before your last merge. Would it be possible to put that as a commit hook?

jburrow commented:
@pyatyispyatil  what you think? I think there is a big introduced in where packaged. JSON is pointing to wing location in first folder... Now tsc is generating in dist/src

jburrow commented:
I have address the issue identified ^ with the commit above  https://github.com/pyatyispyatil/flame-chart-js/pull/44/commits/599429e13d67aec61073c62d6ab5393079188ef4

pyatyispyatil commented on .eslintrc:
```diff
@@ -53,10 +53,16 @@
         // TS rules
         "@typescript-eslint/prefer-optional-chain": "error",
-        "@typescript-eslint/no-unused-vars": "error",
+        "@typescript-eslint/no-unused-vars": [
```
Please remove these changes. We don't want to have any unused variables in our code.)

pyatyispyatil commented on example/src/index.ts:
```diff
@@ -1,13 +1,24 @@
-import { Data, FlameChart } from '../../src';
```
Please revert changes to this file. I'll add your plugin to the example myself.)

jburrow replied:
.... my ability to test is using this code. Would you like to take ownership of this PR and change it to your liking?

jburrow replied:
reverted

pyatyispyatil commented on example/src/test-data.ts:
```diff
@@ -1,4 +1,5 @@
-import { Node, WaterfallIntervals } from '../../src';
```
Please revert changes to this file. You have broken the generation of the flame chart data. You can only add a function that generates data for your plugin.

pyatyispyatil replied:
<img width="1434" alt="image" src="https://user-images.githubusercontent.com/4976306/211203018-65aecb18-3d89-493a-ae3d-a25ca51d5dde.png">

jburrow replied:
I didn't spot that - merge conflict from hell with that - sorry.

jburrow replied:
reverted

pyatyispyatil commented on tsconfig.json:
```diff
@@ -29,15 +29,15 @@
-            "color",
-            "events"
-        ] /* Specify type package names to be included without being referenced in a source file. */,
+        // "types": [
```
Why?

jburrow replied:
It didn't appear to do anything!

pyatyispyatil commented on src/plugins/utils/tree-clusters.spec.ts:
```diff
@@ -1,797 +1,803 @@
-import { describe, it, expect } from '@jest/globals';
```
You need to deal with this file. Why did you comment on everything that was here?

jburrow replied:
I commented it out - cos the tests fail - and it shows the build as failing.

I left a todo for someone to come back and fix.

I think a passing build good.

pyatyispyatil commented on src/plugins/timeseries-plugin.ts:
```diff
@@ -0,0 +1,258 @@
+    }
+
+    setData(data: TimeseriesPoint[]) {
+        this.data = data;
```
How about multiple charts in this plugin?

jburrow replied:
I had no plans to do that.

pyatyispyatil commented on src/flame-chart.ts:
```diff
@@ -90,13 +97,21 @@ export class FlameChart extends FlameChartContainer<FlameChartStyles> {
         }
 
+        if (timeseries) {
+            timeseries.forEach((ts, idx) => {
```
I think this behavior is not expected for our users. Still, it is better to provide for the possibility of adding several charts to one plugin by default. If a user needs multiple plugins, then it is better to use a flame chart container. Otherwise, it will be very confusing, as it seems to me. @Smrtnyk, what do you think?

jburrow replied:
Obviously you know this project best - but the public api that you defined does not expose the internals of the plugins - A user can provide mulitple time-series if they wish...

Materially - It is another chunk of work to refactor the to do this - and i have not planned to do this - as the plugin already meets my use-cases

I would love to chat to you about how i am using it - We have built a custom sampling profiler and associated ui- but the killer feature is the ability to compare multiple profiles with (cpu, memory, other custom attributes too) this has worked almost completly out of the box - this project of your is very very cool!

Smrtnyk replied:
I agree with Nikolay here

pyatyispyatil replied:
It's very cool that you liked the project.)

I think in terms of user experience for the purpose of displaying only one flame chart. And in this case, several graphs at the very bottom after the flame chart look rather strange. For example, in chrome devtools, all charts are built right into the timeframe selector. In our case, of course, we can expand this story and make a separate graph. But at the same time, it should not take up too much space - that's why I suggested making it possible to render several charts at once. After the merge, I'm basically going to slightly improve the api of the plugin that you suggest.

Your goals are most likely related to using the `FlameChartContainer`, which allows you to create complex flame charts. And there you can use the plugins however you like.)

And the `FlameChart` class should, in theory, only allow one chart and its associated information to be displayed.)

Well, for the rest - you can write to me in the telegram. I am available there under the same name.)

pyatyispyatil commented on package.json:
```diff
@@ -2,12 +2,12 @@
     "description": "",
-    "main": "dist/index.js",
-    "types": "dist/index.d.ts",
+    "main": "dist/src/index.js",
```
Why? We don't have src folder in our dist.

jburrow replied:
It is generating a /src/ folder as on this branch - this is fix for that.... 

npm run build:tsc 

And see...

pyatyispyatil replied:
But why is it needed?

pyatyispyatil commented on package.json:
```diff
@@ -2,12 +2,12 @@
     "scripts": {
         "build": "rollup -c",
-        "build:tsc": "tsc",
+        "build:tsc": "tsc -p ./tsconfig.build.json",
```
You broke the build for npm publish.(

jburrow replied:
How do i repro? Can you show me where to look?

pyatyispyatil replied:
We have two builds - one for github pages and one for npm publishing. For the first one, we only provide a dist folder without any nesting, and for github pages, we only create an example folder. You build both.)
It would have been nice if the build had remained untouched, since I don't understand why these changes were made.)

jburrow replied:
I believed there to be an issue - because when i do a npm build:tsc - the files generated do not map main or types in package.json.

I clearly don't understand your project as well as you - but my intentions were that i thought i was making a bug fix. I am sorry

![image](https://user-images.githubusercontent.com/1444266/211219690-27105a2a-4c75-4683-9251-079e1773826d.png)
![image](https://user-images.githubusercontent.com/1444266/211219738-c78a6f8b-c37a-4fb6-98dc-61e2738e091f.png)

pyatyispyatil commented on src/plugins/timeseries-plugin.ts:
```diff
@@ -0,0 +1,258 @@
+            return this.height - value * heightPerValueUnit;
+        };
+
+        const convertDataPointToPointWithIndex = (
```
You should move this function outside of render because the render function needs to be as performant as possible.

jburrow replied:
done

pyatyispyatil commented on src/plugins/timeseries-plugin.ts:
```diff
@@ -0,0 +1,258 @@
+            time: number,
+            value: number
+        ): TimeseriesPointWithIndex => {
+            return [idx, time, value, this.renderEngine.timeToPosition(time), normalizeValue(value)];
```
I didn't see any reason to create so many new arrays. It's better to just process the points in a loop and immediately get the position for each point.
```ts
this.data.forEach(([time, value]) => {
  const pos = this.renderEngine.timeToPosition(time);
  const relativeHeight = this.height - value * heightPerValueUnit;

  this.renderEngine.ctx.lineTo(pos, relativeHeight);
})
```

jburrow replied:
I create 1 array. I perform 2 enumerations.

pyatyispyatil replied:
What I mean is that you can avoid creating an array of datapoints. This array is huge, and if we have, for example, 100k points, we should create an array with 100k arrays with 7 elements. It would take something like 0.5-1ms per frame, and we only have 16ms per frame for all of our calculations and drawing. I think it's too much for a simple chart. Also, you create the same values every time - it's redundant.

jburrow replied:
i will try

pyatyispyatil commented on src/plugins/timeseries-plugin.ts:
```diff
@@ -0,0 +1,258 @@
+
+type TimeseriesPointWithIndex = [idx: number, ts: number, value: number, normalizedTs: number, normalizedValue: number];
+
+enum TimeseriesPointWithIndexX {
```
It a very useless enum. It greatly complicates the code. Better to do without it.

jburrow replied:
The enum was to make the code more readable - to stop hardcode numberic indexes; [0] is less readable than [Enum.timestamp]

If you feel strongly - i can remove it - but the reason i introduced it was i had bugs with my indexes

jburrow replied:
Do you want me to use object instead? therefore not needing the index enum at all?

pyatyispyatil replied:
You can just go the other way - do not create data for yourself on each render, but use the data from `this.data` each time. https://github.com/pyatyispyatil/flame-chart-js/pull/44#discussion_r1064683060

pyatyispyatil commented:
By the way, I did not understand the meaning of the numbers on the left of the graph at all. I didn't specifically look at the code to understand UX. Maybe add at least the designation min max?

pyatyispyatil commented:
> done - i noticed you hadn't run "npm run pretty" before your last merge. Would it be possible to put that as a commit hook?

I think we should change the `npm run lint` command to check for all missing files and readme.md.

jburrow commented:
> By the way, I did not understand the meaning of the numbers on the left of the graph at all. I didn't specifically look at the code to understand UX. Maybe add at least the designation min max?

TBH - I am not sure I like it either you suggestion the word "min value" "max value" post fix the number value `${max} max value`

TBH - I would happily hide the feature behind a setting....

jburrow commented:
@pyatyispyatil - You have given a lot of feedback on this PR - some of this is programming style, some of it is bug's i introduced because i don't know better! [sorry for that]

Given you have strong vision for your project from functionality and code style - would you like to take ownership and just change this code to your liking?

pyatyispyatil commented:
> Given you have strong vision for your project from functionality and code style - would you like to take ownership and just change this code to your liking?
Unfortunately no. If you have already taken up something, then be kind to bring it to the end.)

jburrow commented:
Let 

> > Given you have strong vision for your project from functionality and code style - would you like to take ownership and just change this code to your liking?
> 
> Unfortunately no. If you have already taken up something, then be kind to bring it to the end.)

Understood - I have reverted everything.

Issues I have not addressed;
 - Support multiple time-series in a plugin - is this a deal breaker? if so please tell me now.
 - Refactored internal implementation of render - There are still 2 enumerations, one to find the right data and one to render; My implementation requires this in-order to correct handle the data-points either side of the selected range. There might be a better algo - but i couldn't think of it. I did profile this and it looked ok. Can you give more guidance around performance acceptations criteria?

jburrow commented:
I think there is a problem with the rollup config - possibly due to the use of the builtins() - Which confuses the typechecking not know if it is browser or node? I am not an expert - but here is example

REPO : npm start

![image](https://user-images.githubusercontent.com/1444266/211219883-ad69ca94-61e3-4f8a-bb5a-38bfb49b8442.png)

Smrtnyk commented:
What Nikolay says, I agree with
Writing a plugin should not require changing a bunch of other unrelated stuff
If there is a bug we should first open an issue and try to fix it in a separate PR

Then the new plugin PR should be very clear and only care about the plugin functionality itself

pyatyispyatil commented:
> I think there is a problem with the rollup config - possibly due to the use of the builtins() - Which confuses the typechecking not know if it is browser or node? I am not an expert - but here is example
> 
> REPO : npm start
Well, as I see it, this bug is not reproducible on the master branch.
![image](https://user-images.githubusercontent.com/4976306/211323476-ec4646ea-0aef-4e7e-8b84-e95c6ecef53d.png)

You can try deleting the node_modules folder and running `npm i`. Maybe this will help you.)

jburrow commented:
> > I think there is a problem with the rollup config - possibly due to the use of the builtins() - Which confuses the typechecking not know if it is browser or node? I am not an expert - but here is example
> > REPO : npm start
> 
> Well, as I see it, this bug is not reproducible on the master branch. ![image](https://user-images.githubusercontent.com/4976306/211323476-ec4646ea-0aef-4e7e-8b84-e95c6ecef53d.png)
> 
> You can try deleting the node_modules folder and running `npm i`. Maybe this will help you.)

npm i

FIXED: thank-you and sorry for false report

pyatyispyatil commented:
> Support multiple time-series in a plugin - is this a deal breaker? if so please tell me now.

No, you can only remove redundant creating of several plugins and keep possibility to add only one chart. I will add possibility for creating several charts later.

> Refactored internal implementation of render - There are still 2 enumerations, one to find the right data and one to render; My implementation requires this in-order to correct handle the data-points either side of the selected range. There might be a better algo - but i couldn't think of it. I did profile this and it looked ok. Can you give more guidance around performance acceptations criteria?

I added a comment above about this https://github.com/pyatyispyatil/flame-chart-js/pull/44#discussion_r1064683060. 
You can profile rendering while zooming in/out and scrolling left/right. You should set CPU throttle x4-x6 and try to render 100k-1000k points. I think this plugin should not take more than 0.1ms to render. The requirements may be too high, but still I would like to get as close as possible to this value.

jburrow commented:
![image](https://user-images.githubusercontent.com/1444266/212843595-2e07c982-5b44-4e6b-9faa-78068480e538.png)

jburrow commented:
I believe i have addressed you previous comments - please consider this PR

jburrow commented on src/index.ts:
```diff
@@ -7,6 +7,10 @@ export { WaterfallPlugin, WaterfallPluginSettings } from './plugins/waterfall-pl
 export { TimeseriesPlugin, TimeseriesPluginStyles, TimeseriesPoint } from './plugins/timeseries-plugin';
 export { FlameChart, FlameChart as default } from './flame-chart';
+
+
```
no - ran pretty now

Smrtnyk commented on src/plugins/utils/tree-clusters.spec.ts:
```diff
@@ -77,3 +77,799 @@ describe('tree clusters', () => {
         expect(result).toMatchSnapshot();
     });
 });
+// import { clusterizeFlatTree, flatTree, metaClusterizeFlatTree, reclusterizeClusteredFlatTree } from './tree-clusters';
```
what is this about?

jburrow replied:
resolves - thats result of a merge...

jburrow commented:
Be great if we could close and merge this PR

pyatyispyatil commented on src/plugins/timeframe-selector-plugin.ts:
```diff
@@ -481,3 +474,5 @@ export class TimeframeSelectorPlugin extends UIPlugin<TimeframeSelectorPluginSty
     }
 }
+
+export default TimeframeSelectorPlugin;
```
But we already have a named export here.

pyatyispyatil commented on src/plugins/timeseries-plugin.ts:
```diff
@@ -0,0 +1,270 @@
+        value: number,
+        heightPerValueUnit: number
+    ): TimeseriesPointWithExtra {
+        return {
```
Why are you creating this object? You're just wrapping values (idx, timestamp, value) in an object, which makes no sense.(

pyatyispyatil commented on src/plugins/timeseries-plugin.ts:
```diff
@@ -0,0 +1,270 @@
+                    this.renderEngine.ctx.lineTo(positionStart, prevTPI.normalizedValue);
+                }
+
+                tpi = this.convertDataPointToPointWithIndex(indexStart, time, value, heightPerValueUnit);
```
Just
```ts
const position = this.renderEngine.timeToPosition(time),
const normalizedValue = this.normalizeValue(value, heightPerValueUnit)
```

pyatyispyatil commented on src/plugins/timeseries-plugin.ts:
```diff
@@ -0,0 +1,270 @@
+                if (!foundStart) {
+                    foundStart = true;
+
+                    prevTPI = this.convertDataPointToPointWithIndex(
```
Just
```ts
const position = this.renderEngine.timeToPosition(this.data[indexStart][0]),
const normalizedValue = this.normalizeValue(this.data[indexStart][1], heightPerValueUnit)
```

pyatyispyatil commented:
@jburrow There are a few minor fixes left and I'll be ready to push the merge button.) You did a good job on the pull request, but there are still a couple of things that need to be fixed to improve the code and performance. Please don't think I'm being too picky - graphs and charts are actually very demanding, and creating unnecessary entities (objects, arrays) ends up having a big impact on framerate.

pyatyispyatil commented on src/plugins/timeseries-plugin.ts:
```diff
@@ -0,0 +1,270 @@
+
+                    this.renderEngine.ctx.lineTo(tpi.position, tpi.normalizedValue);
+
+                    this.interactionsEngine.addHitRegion(
```
You can avoid creating hit regions and just calculate the plot value from the mouse position directly in the tooltip's render function. Use a binary search to find the specific nearest point: https://en.wikipedia.org/wiki/Binary_search_algorithm

It should look something like this: https://stackoverflow.com/questions/64737346/binary-search-to-find-closest-float-value-to-target-number

pyatyispyatil replied:
Why is it important:
If there are 10,000 points on the chart, then the engine will need to brute-force check all 10,000 regions for mouse entry. And most importantly, these 10,000 hit areas will be created on every frame. In the case of a binary search, only 13 simple comparison operations are required (base 2 logarithm of 10000)

jburrow replied:
I think i understand what you are saying; But i am not sure which apis to use on the render; 

Please can you describe clearer how you would expect this work.... here is my guess
 - respond to the mouse-move intereaction event; use the mouse.positionX binary search closest data point and set this on the plugin
 - create 1 hit region for the whole of the chart?
 - use the renderToolip() method?
 
 I had a quick  play - but i am not see the renderTooltip method called frequently with this approach

I also tried rendering the tooltip from the move event ... but this had obvious rendering issues

![image](https://user-images.githubusercontent.com/1444266/213258452-22474645-efb2-48a4-b32f-61144de622be.png)

pyatyispyatil replied:
You can only use the `renderTooltip` method and return true from this method. Inside it, you have access to the interactions engine and you can get the position of the mouse from it. I think it should look something like this
```ts
override renderTooltip() {
  const posX = this.interactionsEngine.mouse.x;
  const { time, value } = this.findNearestPoint(pos.x);

  this.renderEngine.renderTooltipFromData(
      [{ text: time }, { text: value }],
      this.interactionsEngine.getGlobalMouse()
  );

  return true;
}
```
The engine should take care of all the work, you just need to describe what exactly the tooltip should show inside the `renderTooltip` method.

If something does not go according to plan, then I think using your code I can figure it out and make changes to the engine, if necessary.)

jburrow replied:
renderTooltip isn't being fired often enough; It does fire - but rarely. Thus the tooltip is stale often. I just put a console.log() inside it to see when it was called.

pyatyispyatil replied:
I checked the engine code better - sorry, but I forgot that you really need to mark the hit region.  I think that for this plugin, you need to make only one region for the entire height and width of the chart.

jburrow replied:
Yup - this does improve things - but it isn't perfect. - I have asked other questions in PR

pyatyispyatil commented:
And tooltip still broken.(
<img width="1438" alt="image" src="https://user-images.githubusercontent.com/4976306/213185954-fc8d3dcd-5ce0-4253-ad27-2cb3e04f90a0.png">

jburrow commented on src/plugins/timeseries-plugin.ts:
```diff
@@ -0,0 +1,215 @@
+
+    override renderTooltip() {
+        const mouse = this.interactionsEngine.getGlobalMouse();
+        //QUESTION: how do i convert mouse.x to position
```
@pyatyispyatil  - How to i convert mouse.x position to timestamp?

@pyatyispyatil - the tooltip kind of works - but it is alittle flaky - sometimes it appears, sometimes it doesn't. there seems to be something not quite right

pyatyispyatil replied:
You can convert pixels to time like this
```ts
this.renderEngine.pixelToTime(this.interactionsEngine.mouse.x) + this.renderEngine.min
```

pyatyispyatil replied:
About glitches, I understand. I made the use of the hover a bit incomprehensible. I will improve this point further, but for now, you need to subscribe to hover and save the region the same way as in other plugins and use it for validation in the tooltip's render method. If there is no region, do not render the tooltip and return false.
<img width="825" alt="image" src="https://user-images.githubusercontent.com/4976306/213403100-4323a21c-6f58-4aae-9ddf-6112b5276915.png">
<img width="644" alt="image" src="https://user-images.githubusercontent.com/4976306/213403168-867019f8-7100-4338-ba72-ee13b47f515d.png">

pyatyispyatil replied:
Yes, I found another hover related bug. If you move the cursor from one of the flame chart blocks directly to your chart, the flame chart tooltip remains. This is an engine bug - I'll deal with it today, you don't have to worry about it.

pyatyispyatil replied:
Fixed in https://github.com/pyatyispyatil/flame-chart-js/commit/b2f5e9d335a89d0e78faa3c425b729bb60d68995

jburrow replied:
great - i found that bug too!

pyatyispyatil replied:
> I think have done everything you asked now;

> Please note - if you tooltip over the flame-chart first the tooltip is broken... but as long as you don't it works just fine...

> I think this is a bug somewhere else in the render engine? or a lack of undertanding on my part.

This bug is on your side. I mentioned it above https://github.com/pyatyispyatil/flame-chart-js/pull/44#discussion_r1080995355
Your tooltip remains because you don't check the hovered region.

https://github.com/pyatyispyatil/flame-chart-js/pull/44#discussion_r1081142303 - this message about another one.)

jburrow commented on src/plugins/timeseries-plugin.ts:
```diff
@@ -0,0 +1,215 @@
+        const heightPerValueUnit =
+            (this.height - this.styles.padding) / Math.max(this.summary.max - this.summary.min, 1);
+
+        this.interactionsEngine.addHitRegion(RegionTypes.CLUSTER, 'not-used', 0, 0, width, this.height);
```
This seemed to help - but 100% sure if it does or not

pyatyispyatil replied:
Yes, it works. But you need to add a new enum for the plugin.)

jburrow commented:
I think have done everything you asked now; 

Please note - if you tooltip over the flame-chart first the tooltip is broken... but as long as you don't it works just fine...

I think this is a bug somewhere else in the render engine? or a lack of undertanding on my part.

jburrow commented:
Please can you run "npm run pretty" on master :)

jburrow commented:
i am at a loss now - just merged your latest changes which re-writes the example app :( - i could support timeseries there but tbh - given you have a strong opinion on how this should work - maybe you should take it from here?

pyatyispyatil commented:
> Please can you run "npm run pretty" on master :)

Why is this needed? On the master, everything is fine with linting, but at the same time, your branch contains many problems.

pyatyispyatil commented:
> i am at a loss now - just merged your latest changes which re-writes the example app :( - i could support timeseries there but tbh - given you have a strong opinion on how this should work - maybe you should take it from here?

Okay, I'll add.)

jburrow commented:
> > Please can you run "npm run pretty" on master :)
> 
> Why is this needed? On the master, everything is fine with linting, but at the same time, your branch contains many problems.

I was referring to the fact the configs in the route of your project haven't be formatted - anyhow - leaving it to you.

jburrow commented:
I have removed everthing but the plugin - as you can choose how to use - in the default|customized route etc.

Please let me know if you can merge this or not.

pyatyispyatil commented:
> I have removed everthing but the plugin - as you can choose how to use - in the default|customized route etc.
> 
> Please let me know if you can merge this or not.

I can’t merge yet, because  you still need to fix the bug with the tooltip from this thread https://github.com/pyatyispyatil/flame-chart-js/pull/44#discussion_r1083460210

pyatyispyatil commented:
And you have linting problems that are marked on CI.

jburrow commented:
> And you have linting problems that are marked on CI.


I don't understand - I have ran npm run pretty - it doesn't update test-data.ts - ... it does update 5 files in teh route of the project thought.

How do i repro what vercel is doing....

jburrow commented:
> > I have removed everthing but the plugin - as you can choose how to use - in the default|customized route etc.
> > Please let me know if you can merge this or not.
> 
> I can’t merge yet, because you still need to fix the bug with the tooltip from this thread [#44 (comment)](https://github.com/pyatyispyatil/flame-chart-js/pull/44#discussion_r1083460210)

I have no development harness now - because it was removed when you re-wrote the app to be typescript.
I would need to re-write the a development harness - but i feel that whatever i do won't be aligned to your expectations.

What do you recommend i do?

pyatyispyatil commented:
> > And you have linting problems that are marked on CI.
> 
> I don't understand - I have ran npm run pretty - it doesn't update test-data.ts - ... it does update 5 files in teh route of the project thought.
> 
> How do i repro what vercel is doing....

`npm run lint:fix`

pyatyispyatil commented:
> > > I have removed everthing but the plugin - as you can choose how to use - in the default|customized route etc.
> > > Please let me know if you can merge this or not.
> > 
> > 
> > I can’t merge yet, because you still need to fix the bug with the tooltip from this thread [#44 (comment)](https://github.com/pyatyispyatil/flame-chart-js/pull/44#discussion_r1083460210)
> 
> I have no development harness now - because it was removed when you re-wrote the app to be react. I would need to re-write the a development harness - but i feel that whatever i do won't be aligned to your expectations.
> 
> What do you recommend i do?

Dude, I just want the code in the project to work well (without major bugs), and that it does not lead to unnecessary performance costs and passes all the checks in the pipeline. This package is already used by several projects and I would not like to ship something low-quality to them.)

Many of the comments were simply unfixed by you, which is why this pull request stuck in this status for so long.(

You can add initialization of your plugin to the new example to check how it works. And after that fix the bug that was described in the comments.
